examples/Example7/example7-adept.cc

```diff
@@ -0,0 +1,193 @@
+  if (world == nullptr) {
+    std::cerr << "GeoManager world volume is nullptr" << std::endl;
+    return 3;
+  }
+#ifdef VERBOSE
+  std::cout << "World (ID " << world->id() << ")" << std::endl;
+  PrintDaughters(world);
+#endif
+
+  // G4HepEmData
+  // Input: file path (here from vgdml::Middleware)
+  // Output: G4HepEmState pointer (owned)
+  const auto& userInfo = middleWare->GetUserInfo();
+  vgdml::Auxiliary g4hepemFileAux = find_by_type(userInfo, "g4hepemfile");
+
+  if (g4hepemFileAux == vgdml::Auxiliary{}) {
+    std::cerr << "GDML file '" << gdml_file << "' has no 'g4hepemfile' user info auxiliary" << std::endl;
+    return 4;
+  }
+
+  std::string g4hepem_file = g4hepemFileAux.GetValue();
+  std::ifstream jsonIS{ g4hepem_file.c_str() };
+
+  G4HepEmState* inState = G4HepEmStateFromJson(jsonIS);
+  if(inState == nullptr)
+  {
+    std::cerr << "Failed to read G4HepEmState from " << g4hepem_file << std::endl;
+    return 4;
+  }
+
+  // Connect up VecGeom _physical_ volume ids to G4HepEMData material cut couple indexes
+  // via Geant4 material cut couple index supplied as a volume auxiliary in the GDML
+  // Inputs: vecgeom::GeoManager, vgdml::Middleware, G4HepEmData
+  // Output: Array of ints. Array indices are VecGeom PV id numbers, value at that index is G4HepEmData MCC index
+  // Visting all PVs walks the tree, so we get multiple copies in the vector. Thus reduced to a
+  // set of pointers afterwards to uniq the list.
+  std::vector<vecgeom::VPlacedVolume*> tmpPlacedVolumes;
+  vecgeom::GeoManager::Instance().getAllPlacedVolumes(tmpPlacedVolumes);
+  std::set<vecgeom::VPlacedVolume*> allPlacedVolumes{tmpPlacedVolumes.begin(), tmpPlacedVolumes.end()};
+
+  const auto NumVolumes = allPlacedVolumes.size();
+  int* MCCIndex = new int[NumVolumes];
+
+  const auto& lvAuxInfo = middleWare->GetVolumeAuxiliaryInfo();
+
+  G4HepEmData* inData = inState->fData;
+
+  for (const vecgeom::VPlacedVolume* pv : allPlacedVolumes) {
+    const auto lvId = pv->GetLogicalVolume()->id();
+
+    auto lvAuxForThisLV = lvAuxInfo.find(lvId);
+    if(lvAuxForThisLV == lvAuxInfo.end()) {
+      std::cerr << "Logical volume " << lvId << " has no auxiliary data" << std::endl;
+      return 5;
+    }
+
+    // Auxiliary to find has type "g4matcutcouple"
+    vgdml::Auxiliary mccAux = find_by_type(lvAuxForThisLV->second, "g4matcutcouple");
+    if (mccAux == vgdml::Auxiliary{}) {
+      std::cerr << "Logical volume " << lvId << " has no auxiliary of type 'g4matcutcouple'" << std::endl;
+      return 5;
+    }
+
+    // Value is of auxiliary is Geant4 mcc index, so use G4HepEm to get index conversion
+    int geant4MCCIdx = std::stoi(mccAux.GetValue());
+    MCCIndex[pv->id()] = inData->fTheMatCutData->fG4MCIndexToHepEmMCIndex[geant4MCCIdx];
+  }
+
+  // Place particles between the world boundary and the calorimeter.
+  double startX = (-0.5 * fWorldSizeX) + 1.0*copcore::units::mm;
```

Comment:
I was wondering if we could avoid sharing a common geometry header with constants by injecting auxiliary key-value pairs for the constants we need in the gdml file, which would become ideally the single source for the setup configuration in AdePT. We could have:
```
<auxiliary auxtype="Generator" auxvalue="gun">
  <auxiliary auxtyype="xgun" auxunit="mm" auxvalue="startX"/>  // well, what startX value is
  <auxiliary auxtyype="ygun" auxunit="mm" auxvalue="startY"/>
  ...
</auxiliary>
```
For other generators, the content could be e.g. the file name with hepmc3 particles, .... We can actually pass also other generic constants (for which the meaning is known apriori) via this mechanism if we need. This would allow having the AdePT app generic, dependent only on the gdml content. And if you want to run it with some different parameters (e.g. changing generator parameters), you could just modify the gdml. Since gdml also allows having parts written in different files, we could also keep the configuration part in a separate file included from the main gdml. I'm just trying to figure out a way to avoid writing a different app for each setup we want to test...

Reply:
I don't think it makes sense to pass generator settings via the geometry description; we should really put our efforts in having the connection to Geant4 instead of artificially pumping the standalone demonstrators. If anything, we could move the geometry such that the generator is at `(0, 0, 0)`...

Reply:
Generator settings are usually input program parameters and they could stay this way. However, hardcoding numbers of layers, distances, and other geometry-specific parameters should not be needed, I mean I would avoid computing gun position based on such data. No, I don't think it is a good idea to move the geometry.

Reply:
But they are *not* part of the geometry description.

Reply:
auxiliary info does not necessarily need to be part of the geometry description, for example, region production cuts are user-defined properties that connect to the geometry. It could be done the other way around and have a setup file containing this info and linking to the gdml file. The point is that to have the application reusable, it should not contain in the compiled code recipes that are setup dependent. I realize however that this may be a long shot in what concerns for example scoring, which is very much setup-specific, so maybe we give it another thought after we merge this...

Reply:
Anticipating the fact that we will have input particles from LHCb in a file (or passed from Geant4 through the fast simulation  hook) we should implement some simple 'primary particle' interface in AdePT which allows to inject particles from any external source. In other words, there should be a container of primary particles, that we populate the way we want and then AdePT does 'get next primary particle' from that container.

Reply:
I've changed this so that the "startX" point is calculated from the world volume extent. Not perfect, but better than nothing for this example. @WitekPokorski's suggestion is the right long term solution overall.